src/NSL/MarkovChain/RadialHMC.tpp

```diff
@@ -0,0 +1,342 @@
+        Type proposal_S = this->action_(proposal_config);
+
+        // compute the Hamiltonian H = p^2/2 + S
+        // Starting point of the trajectory
+        Type starting_H = state.actionValue;
+        for( const auto& [key,field]: momentum){
+            starting_H += 0.5*(field * field).sum();
+        }
+
+        
+        // End point of the trajectory i.e. proposal
+        Type proposal_H = proposal_S;
+        for( const auto& [key,field]: proposal_momentum){
+            proposal_H += 0.5*(field * field).sum();
+        }
+
+        // We always assume real part of the action, i.e. automatic reweighting
+        // for complex actions!
+        NSL::RealTypeOf<Type> acceptanceProb = NSL::LinAlg::exp( NSL::real(starting_H - proposal_H) );
+
+        // accept reject
+	    if ( r_.rand()[0] <= acceptanceProb ){
+            return NSL::MCMC::MarkovState<Type>{
+                proposal_config,
+                proposal_S,
+                acceptanceProb,
+                state.markovTime+1,
+                true
+                /*For this algorithm there are no weights to be added*/
+            };
+        } else {
+            return NSL::MCMC::MarkovState<Type>(
+                state.configuration,
+                state.actionValue,
+                state.acceptanceProbability,
+                state.markovTime+1,
+                false
+            );
+        }
+    }
+
+    //! Implementation of the HMC radial update
+    // Only implemented for one field component at every lattice site!
+    template<NSL::Concept::isNumber Type>
+    NSL::MCMC::MarkovState<Type> radialgenerate_(const NSL::MCMC::MarkovState<Type> & state, NSL::RealTypeOf<Type> volume, NSL::RealTypeOf<Type> radialScale){
+        NSL::RealTypeOf<Type> rscale_ = rscaleTensor_.lognormal(0., radialScale)[0];
+        // compute the Action
+        NSL::Configuration<Type> proposal_config (state.configuration,true);
+        proposal_config = static_cast<Type>(rscale_) * proposal_config; 
+        Type proposal_S = this->action_(proposal_config);
+
+        // Starting point of the trajectory
+        Type starting_S = state.actionValue;
+
+        // We always assume real part of the action, i.e. automatic reweighting
+        // for complex actions!
+        NSL::RealTypeOf<Type> acceptanceProb = NSL::LinAlg::exp( NSL::real(starting_S - proposal_S) + volume * NSL::LinAlg::log(rscale_)) ;
+
+        // accept reject 
+	    if ( r_.rand()[0] <= acceptanceProb ){
+            return NSL::MCMC::MarkovState<Type>{
+                proposal_config,
+                proposal_S,
+                acceptanceProb,
+                state.markovTime+1,
+                true
+                /*For this algorithm there are no weights to be added*/
+            };
+        } else {
+            return NSL::MCMC::MarkovState<Type>(
+                state.configuration,
+                state.actionValue,
+                state.acceptanceProbability,
```

Comment:
If you want to use the autotuner this should be rejected with the current acceptance probability